client/lib/signup/analytics.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * Internal dependencies
+ */
+import analytics from 'lib/analytics';
+import {
+	adTrackSignupStart,
+	adTrackSignupComplete,
+	adTrackRegistration,
+} from 'lib/analytics/ad-tracking';
+
+export function recordSignupStart( flow, ref ) {
+	// Tracks
+	analytics.tracks.recordEvent( 'calypso_signup_start', { flow, ref } );
+	// Google Analytics
+	analytics.ga.recordEvent( 'Signup', 'calypso_signup_start' );
+	// Marketing
+	adTrackSignupStart( flow );
+}
+
+export function recordSignupComplete(
+	{ flow, isNewUser, isNewSite, hasCartItems, isNew7DUserSite },
+	now
+) {
+	if ( ! now ) {
+		// Delay using the analytics localStorage queue.
+		return analytics.queue.add(
+			'recordSignupComplete',
+			{ flow, isNewUser, isNewSite, hasCartItems, isNew7DUserSite },
+			true
+		);
+	}
+
+	// Tracks
+	analytics.tracks.recordEvent( 'calypso_signup_complete', {
```

Comment:
@jsnajdr Thanks for tackling this. The improvements are much needed and your work is certainly appreciated. I also appreciate that you're involving marketing devs in this decision. Really nice, thanks!

You wrote...

> I'm trying to move out the section specific code from lib/analytics
> to better places like client/signup or client/checkout.

@MicBosi  writes...

> I'm a big fan of restructuring the current analytics library in modules or even sub-libraries but not of spreading the analytics concerns across Calypso's code base into non-analytics related modules.

I just want to add a ➕**1** on this because the `calypso_signup_start` and `calypso_signup_complete` events, among others (@MicBosi can provide a more complete list) are in the critical path where paid search, display ads, and Google Optimize A/B tests all get their data. Sometimes using Tracks, sometimes using GA events, sometimes using ad tracking. As a result, Marketing devs are frequently asked to audit, modify, or otherwise update these.

If we can find a way to keep all of mission-critical events in one place (maybe we group them together and make these an exception) that would be great. Ideally, the Tracks, GA, and ad tracking calls are all together so that we can be absolutely sure that changing props or adding something new is going to cover us in all cases.

What would be bad is if we changed the props in this block of code, and only later we discover there was another instance of this `calypso_signup_complete` event being fired. Or, as happened recently, a suffixed variation.

What do you think?

Reply:
Also, I realize searching the codebase is easy enough when one knows the event they're looking for. I'm thinking about onboarding & discoverability for marketing devs when I imagine these events being together in one place.